### DISCUSSION TITLE
with-apollo example SSR question

### DISCUSSION THREAD

**Discussion by mondoadams** [vercel/next.js#13848] [Help] [answered]
I see the `with-apollo` [example](https://github.com/vercel/next.js/tree/canary/examples/with-apollo) has been updated with the new data fetching methods.

My question is how would I do server-side rendering with the new example? The previous version of this example was setup to do SSR on all Apollo queries, whereas now it has a SSG example and a client-side example.

The SSG page looks like this:

```
export async function getStaticProps() {
  const apolloClient = initializeApollo()

  await apolloClient.query({
    query: ALL_POSTS_QUERY,
    variables: allPostsQueryVars,
  })

  return {
    props: {
      initialApolloState: apolloClient.cache.extract(),
    },
    unstable_revalidate: 1,
  }
}
```

Would I be right to do this then for SSR:
```
export async function getServerSideProps() {
  const apolloClient = initializeApollo()

  await apolloClient.query({
    query: ALL_POSTS_QUERY,
    variables: allPostsQueryVars,
  })

  return {
    props: {
      initialApolloState: apolloClient.cache.extract(),
    },
  }
}
```
It seems to work by quickly testing but wondering if I'm missing something obvious here, and there may be another intended method?

**Comment by pouyajabbarisani** [comment]
I have same question.

**Reply by alex-r89** [reply]
I also have a similar question (it would be nice if the `with-apollo` example has an SSR page!).

How do you go about passing the data from the `getServerSideProps` call in this example? Do you have the queries on the client too which then  get called at some point?

The reason I am asking is because I am noticing that my SSR pages take ages to load (rendering starts at a staggering 4.6 seconds). I am rending like so currently:

```js
const PaginatedPage = ({ photos, total }) => {
  return (
    <main>
      <Header />
      <PhotoGrid photos={photos} />
      <Pagination total={total} />
    </main>
  )
}

export const getServerSideProps = async ({ query }) => {
  const page = parseInt(query.index)
  const client = await initApolloClient({})

  const { data: photoData } = await client.query({
    query: getPhotosQuery,
    variables: { input: { page } }
  })

  const { data: total } = await client.query({
    query: getTotalCount
  })

  return { props: { photos: photoData?.getPhotos, total: total?.totalCount } }
}
```

I am wondering if I can optimise this some how, because a 4.6 second Start Render is terrible!

**Comment by dhazeley** [comment, marked as answer]
> I see the `with-apollo` [example](https://github.com/vercel/next.js/tree/canary/examples/with-apollo) has been updated with the new data fetching methods.
> 
> My question is how would I do server-side rendering with the new example? The previous version of this example was setup to do SSR on all Apollo queries, whereas now it has a SSG example and a client-side example.
> 
> The SSG page looks like this:
> 
> ```
> export async function getStaticProps() {
>   const apolloClient = initializeApollo()
> 
>   await apolloClient.query({
>     query: ALL_POSTS_QUERY,
>     variables: allPostsQueryVars,
>   })
> 
>   return {
>     props: {
>       initialApolloState: apolloClient.cache.extract(),
>     },
>     unstable_revalidate: 1,
>   }
> }
> ```
> 
> Would I be right to do this then for SSR:
> 
> ```
> export async function getServerSideProps() {
>   const apolloClient = initializeApollo()
> 
>   await apolloClient.query({
>     query: ALL_POSTS_QUERY,
>     variables: allPostsQueryVars,
>   })
> 
>   return {
>     props: {
>       initialApolloState: apolloClient.cache.extract(),
>     },
>   }
> }
> ```
> 
> It seems to work by quickly testing but wondering if I'm missing something obvious here, and there may be another intended method?

This would be the correct way to do it. I don't see any reason to complicate things beyond this.

If you wanted an alternative approach, you could use the apollo client as you are now, but pass the data from the query in rather than the cache from the client. Then in your page component, you can use the data directly from the props, rather than making another query in the browser to pull the data from the cache.
This approach would mean you can effectively remove apollo client from your website (assuming you only need to run graphql queries on the server)

**Reply by dhazeley** [reply]
Following up on my alternative approach, if you want to avoid prop drilling, I'd recommend using React Context on the `_app.js` component and setting the Provider value to the graphql data in the `pageProps`

**Reply by sromexs** [reply]
this is the answer
![image](https://user-images.githubusercontent.com/48912836/84452529-4823b800-ac6b-11ea-90c6-62dad9f618c6.png)

**Reply by dhazeley** [reply]
No offense, I don't see how this approach is better than the new `with-apollo` example. This seems like a more minimal and less flexible approach with no upsides.

**Reply by mondoadams** [reply]
@dhazeley thank you, just needed to sense-check myself. I'm actually using a hybrid of CSR pages and SSR so passing the cache to the client works fine for me.

**Reply by alex-r89** [reply]
@dhazeley thanks for the solution, I was wondering this too.

A question I have for both @mondoadams and @dhazeley is; why would you want to do SSR over SSG here? I understand the difference between the two, but don't understand how SSG does not satisfy your requirements?

Any explanation would be really appreciated!

**Comment by seanulacra** [comment]
Also curious about this, noticing that I'm seeing DataSources in the Resolver cause the SchemaLink to fail when initializing on the pages as well

**Reply by chris-pica** [reply]
@seang Seeing the same, did you ever find a solution? Specifically this error because data sources are not initialised in the resolver context in the SSR condition:
`Error: Cannot destructure property 'dataSources' of 'undefined' as it is undefined.`

This is just using the example at https://github.com/vercel/next.js/tree/canary/examples/api-routes-apollo-server-and-client and then adding a RESTDataSource.

**Comment by samfromaway** [comment]
I had the same issue. I have been searching for hours, now to find the solution here in the discussion tab. I think it would be good to have an example page in the repo folder "with-apollo" as well. If that's ok I would even add a PR with the solution displayed here. Let me know and I'll add it.

**Comment by lauriskuznecovs** [comment]
Just wanted to point out some use-case is not covered or at least I cannot get it working with latest Next.j v10.

https://github.com/vercel/next.js/discussions/11957#discussioncomment-552260
